File: sdks/python/agentsdk/crashdump.py
"""CrashDump — structured crash data with atomic write.

Represents the data written to disk when the application crashes due to a
signal or panic. Uses an atomic write-to-tmp-then-rename pattern to avoid
partial files on crash.

The filename format is ``crash-YYYYMMDD-HHMMSS.json`` for easy chronological
sorting.
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Any, Dict, Optional

from agentsdk.sandbox import Sandbox


@dataclass
class CrashDump:
    """Structured record of an application crash for post-mortem debugging."""

    timestamp: str
    app_name: str
    app_version: str
    crash_type: str
    stack_trace: str
    flight_context: Dict[str, Any] = field(default_factory=dict)
    trace_id: str = ""
    signal: str = ""
    panic_value: str = ""

    def to_dict(self) -> Dict[str, Any]:
        """Serialize to a dict, omitting empty/None fields.

        Matches Go SDK's ``omitempty`` semantics — ``trace_id``, ``signal``,
        and ``panic_value`` are excluded when empty.
        """
        d: Dict[str, Any] = {
            "timestamp": self.timestamp,
            "app_name": self.app_name,
            "app_version": self.app_version,
            "crash_type": self.crash_type,
            "stack_trace": self.stack_trace,
            "flight_context": self.flight_context,
        }
        if self.trace_id:
            d["trace_id"] = self.trace_id
        if self.signal:
            d["signal"] = self.signal
        if self.panic_value:
            d["panic_value"] = self.panic_value
        return d


def write_crash_dump(sandbox: Sandbox, dump: CrashDump) -> str:
    """Atomically write a CrashDump as formatted JSON to the crash_dumps/ directory.

    The write uses a write-to-tmp-then-rename pattern (via ``os.replace()``)
    to avoid partial files on crash.  The ``crash_dumps/`` directory is
    created automatically if it does not exist.

    Args:
        sandbox: The application Sandbox providing the crash_dumps directory.
        dump: The CrashDump to write.

    Returns:
        The path of the written crash dump file.

    Raises:
        OSError: If the directory cannot be created, the temp file cannot be
            written, or the atomic rename fails.
        TypeError: If the CrashDump cannot be serialized to JSON.
    """
    data = json.dumps(dump.to_dict(), indent=2)

    crash_dir = Path(sandbox.crash_dumps_dir)
    crash_dir.mkdir(parents=True, exist_ok=True)

    filename = f"crash-{datetime.now().strftime('%Y%m%d-%H%M%S')}.json"
    final_path = crash_dir / filename
    tmp_path = str(final_path) + ".tmp"

    with open(tmp_path, "w", encoding="utf-8") as f:
        f.write(data)

    os.replace(tmp_path, str(final_path))

    return str(final_path)

File: sdks/go/crashdump.go
package agentsdk

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"time"
)

// CrashDump represents the structured JSON data written to disk when the
// application crashes due to a signal or panic. Each field captures a
// different dimension of the crash scene for post-mortem debugging.
type CrashDump struct {
	// Timestamp is the crash time in RFC3339 format.
	Timestamp string `json:"timestamp"`

	// AppName is the name of the application that crashed.
	AppName string `json:"app_name"`

	// AppVersion is the version of the application that crashed.
	AppVersion string `json:"app_version"`

	// TraceID is the distributed trace ID active at crash time, if any.
	TraceID string `json:"trace_id,omitempty"`

	// CrashType is either "signal" or "panic".
	CrashType string `json:"crash_type"`

	// Signal is the OS signal name (e.g. "SIGTERM", "SIGINT").
	// Empty for panic-type crashes.
	Signal string `json:"signal,omitempty"`

	// PanicValue is the string representation of the panic value.
	// Empty for signal-type crashes.
	PanicValue string `json:"panic_value,omitempty"`

	// StackTrace contains the goroutine stack trace at crash time.
	StackTrace string `json:"stack_trace"`

	// FlightContext holds a snapshot of in-flight state at crash time.
	FlightContext map[string]interface{} `json:"flight_context"`
}

// WriteCrashDump atomically writes a CrashDump as formatted JSON to the
// crash_dumps/ directory managed by the Sandbox.
//
// The write uses the same write-to-tmp-then-rename pattern as ConfigManager
// to avoid partial files on crash. The filename format is crash-YYYYMMDD-HHMMSS.json
// for easy chronological sorting.
//
// If the crash_dumps/ directory does not exist, it is created via os.MkdirAll.
// Write failures are non-fatal — the function returns the error but does not
// panic, allowing callers to log and continue their shutdown sequence.
func WriteCrashDump(sandbox *Sandbox, dump *CrashDump) error {
	data, err := json.MarshalIndent(dump, "", "  ")
	if err != nil {
		return fmt.Errorf("crashdump: marshal: %w", err)
	}

	dir := sandbox.CrashDumpsDir()
	if err := os.MkdirAll(dir, 0755); err != nil {
		return fmt.Errorf("crashdump: create dir %q: %w", dir, err)
	}

	filename := fmt.Sprintf("crash-%s.json", time.Now().Format("20060102-150405"))
	finalPath := filepath.Join(dir, filename)
	tmpPath := finalPath + ".tmp"

	if err := os.WriteFile(tmpPath, data, 0644); err != nil {
		return fmt.Errorf("crashdump: write tmp %q: %w", tmpPath, err)
	}

	if err := os.Rename(tmpPath, finalPath); err != nil {
		return fmt.Errorf("crashdump: rename %q → %q: %w", tmpPath, finalPath, err)
	}

	return nil
}
